torch/lib/c10d/TCPStore.cpp

```diff
@@ -39,15 +39,69 @@ enum class WatchResponseType : uint8_t { KEY_UPDATED };
 // TCPStoreListener class methods
 ListenThread::ListenThread(int listenSocket)
   : storeListenSocket_(listenSocket) {
+  // init stop signal
+  if (pipe(controlPipeFd_.data()) == -1) {
+    throw std::runtime_error(
+        "Failed to create the control pipe to start the "
+        "TCPStoreListener thread");
+  }
+  listenerThread_ = std::thread(&ListenThread::run, this);
 }
 
 ListenThread::~ListenThread() {
+  if (controlPipeFd_[1] != -1) {
+    // close the write end of the pipe
+    ::close(controlPipeFd_[1]);
+    controlPipeFd_[1] = -1;
+  }
+
+  listenerThread_.join();
+
+  for (auto fd : controlPipeFd_) {
+    if (fd != -1) {
+      ::close(fd);
+    }
+  }
 }
 
 void ListenThread::run() {
+  std::vector<struct pollfd> fds;
+  tcputil::addPollfd(fds, controlPipeFd_[0], POLLHUP);
+  tcputil::addPollfd(fds, storeListenSocket_, POLLIN);
+
+  while (1) {
+    SYSCHECK_ERR_RETURN_NEG1(::poll(fds.data(), fds.size(), -1));
```

Comment:
On the server side, doesn't this poll interfere with the poll in TCPStoreDaemon?

Reply:
We now have to create a separate thread on the client to handle responses asynchronously from the master, this is running in that separate thread. I updated the first PR with more details.